### PR TITLE
feat(transform): Add ToJson string conversion

### PR DESCRIPTION
**Pull request by MisterMX**
### Description of your changes

This adds a new `ToJson` conversion option to the string transform which allows converting the input value into its raw JSON representation:

```yaml
- type: string
  string:
     type: Convert
     convert: ToJson
```

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Manually and unit tests.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Review by nullable-eth — APPROVED** [review]
Looks great!

File: apis/apiextensions/v1/composition_transforms.go [hunk 1]
```diff
@@ -61,6 +61,7 @@ const (
 	errStringConvertTypeFailed          = "type %s is not supported for string convert"
 
 	errDecodeString = "string is not valid base64"
+	errMarshalJson  = "failed to marshal value to JSON"
```
Comment (turkenh) [inline apis/apiextensions/v1/composition_transforms.go:64]:
```suggestion
	errMarshalJson  = "cannot marshal value to JSON"
```

nitpick: just to be consistent with other error messages.

**Review by turkenh — APPROVED** [review]

**Comment by ezgidemirel** [comment]
Hi @MisterMX, thanks for the PR! Would you mind sharing snippets of your test manifests?

**Comment by MisterMX** [comment]
@ezgidemirel here is a working example:

`composition.yaml`
```yaml
---
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: test-string-tojson
spec:
  compositeTypeRef:
    apiVersion: example.org/v1alpha1
    kind: XSecret
  resources:
    - name: secret
      base:
        apiVersion: secretsmanager.aws.crossplane.io/v1beta1
        kind: Secret
        metadata:
          name: # patched
        spec:
          forProvider:
            region: eu-central-1
            stringSecretRef:
              name: # patched
              namespace: # patched
              key: # patched
            resourcePolicy: # patched
          providerConfigRef:
            name: example
      patches:
        # Policy
        - fromFieldPath: "spec.parameters.resourcePolicy"
          toFieldPath: "spec.forProvider.resourcePolicy"
          transforms:
            - type: string
              string:
                type: Convert
                convert: ToJson
        # SecretRef
        - fromFieldPath: "spec.parameters.secretRef.name"
          toFieldPath: "spec.forProvider.stringSecretRef.name"
        - fromFieldPath: "spec.parameters.secretRef.key"
          toFieldPath: "spec.forProvider.stringSecretRef.key"
        - fromFieldPath: "metadata.labels[crossplane.io/claim-namespace]"
          toFieldPath: "spec.forProvider.stringSecretRef.namespace"
```

`xrd.yaml`
```yaml
---
apiVersion: apiextensions.crossplane.io/v1
kind: CompositeResourceDefinition
metadata:
  name: xsecrets.example.org
spec:
  group: example.org
  claimNames:
    kind: Secret
    plural: secrets
  names:
    kind: XSecret
    plural: xsecrets
  versions:
    - name: v1alpha1
      served: true
      referenceable: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec:
              type: object
              properties:
                parameters:
                  type: object
                  properties:
                    resourcePolicy:
                      type: object
                      description: AWS Secret policy object.
                      x-kubernetes-preserve-unknown-fields: true
                    secretRef:
                      type: object
                      description: The Kubernetes Secret whose data will be sent as string to AWS.
                        The secret needs to exist in the same namespace as this claim.
                      properties:
                        key:
                          type: string
                          description: Optional key whose value should be uploaded. If not set, the
                            whole map in the Secret data will be used.
                        name:
                          type: string
                          description: Name of the Kubernetes Secret resource in the same namespace
                            as this claim.
                      required:
                        - name
              required:
                - parameters
```

`xrc.yaml`
```yaml
---
apiVersion: v1
kind: Secret
metadata:
  name: example-secret
  namespace: default
type: Opaque
stringData:
  password: passw0rd
---
apiVersion: example.org/v1alpha1
kind: Secret
metadata:
  name: example-secret
  namespace: default
spec:
  parameters:
    secretRef:
      name: example-secret
    resourcePolicy:
      {
        "Version": "2012-10-17",
        "Statement": [
          {
            "Effect": "Allow",
            "Principal": {
              "Federated": [
                "*"
              ]
            },
            "Action": "secretsmanager:GetSecretValue",
            "Resource": "*"
          }
        ]
      }
```

**Comment by MisterMX** [comment]
I can't get the checks to succeed. It seems like an internal error in the pipeline. @ezgidemirel do you have any ideas?

File: apis/apiextensions/v1/composition_transforms.go [hunk 2]
```diff
@@ -61,6 +61,7 @@ const (
 	errStringConvertTypeFailed          = "type %s is not supported for string convert"
 
 	errDecodeString = "string is not valid base64"
+	errMarshalJson  = "cannot to marshal value to JSON"
```
Comment (hasheddan) [inline apis/apiextensions/v1/composition_transforms.go:64]: @MisterMX looks like linter doesn't like this naming:
```
  Error: var-naming: const errMarshalJson should be errMarshalJSON (revive)
```

Comment (MisterMX) [reply apis/apiextensions/v1/composition_transforms.go]: Thanks for the hint. I fixed it.

**Comment by apj0nes** [comment]
@hasheddan do you know if this feature will go into the next crossplane release?

**Review by hasheddan — APPROVED** [review]
This is awesome -- thanks @MisterMX! I am +1 on inclusion in v1.11.0 release 👍🏻

**Comment by bobh66** [comment]
If we could combine this with a FromJSON transform we could patch complex objects between resources in a composition.  Use ToJSON to patch a list/map into a string annotation, and then FromJSON to patch that annotation into another resource.  Maybe in a separate PR but it would be really useful.

File: docs/reference/composition.md [hunk 3]
```diff
@@ -637,6 +637,13 @@ Currently only `multiply` is supported.
      type: Convert
      convert: FromBase64
 
+# If the value of the 'from' field is not nil, the value of the 'to' field will be
```
Comment (turkenh) [inline docs/reference/composition.md:640]: @plumbis This PR has this doc change. Having this updated here is enough, or do we need a separate PR in the new [docs repo](https://github.com/crossplane/docs)?
